### PR TITLE
pybind11 CMake module refactor

### PR DESCRIPTION
**Pull request by michdolan**
This PR reworks the `Findpybind11.cmake` module to better handle version detection when possible, and warn when it is not. Three methods for locating pybind11 are now used, in this order:

1. If the user specified `-Dpybind11_DIRS=<path>` in the cmake command, try to use that first, and only search that path, ignoring typical system and cmake paths that would be searched otherwise.
2. Try to use a pybind11 CMake config, which should have been installed on the system with pybind11
3. If pybind11 was installed from PyPI using pip, get the path and version from a python process

I've tested these scenarios on my machine and now the pybind11 version number is either discovered, or it is assumed the user knows what they are doing if they override the pybind11 path, and warns them about the version assumption (there doesn't appear to be any version indication in the headers).

This PR should solve #998 

Signed-off-by: Michael Dolan <michdolan@gmail.com>

### CODE REVIEW

File: share/cmake/modules/Findpybind11.cmake [hunk 1]
```diff
@@ -22,28 +23,88 @@ endif()
...
-            pybind11/pybind11.h
-        HINTS
-            ${_pybind11_SEARCH_DIRS}
-        PATH_SUFFIXES
-            include
-            pybind11/include
-    )
+        if(EXISTS "${pybind11_INCLUDE_DIR}")
+            # Assume user-provided version is OK.
+            # TODO: Can the pybind11 version be inspected from a header?
```
Comment (hodoulp) [inline share/cmake/modules/Findpybind11.cmake:42]: Yes. The code could parse the file `ext/dist/include/pybind11/detail/common.h` to find the version:
```
#define PYBIND11_VERSION_MAJOR 2
#define PYBIND11_VERSION_MINOR 4
#define PYBIND11_VERSION_PATCH 3
```

File: share/cmake/modules/Findpybind11.cmake [hunk 2]
```diff
@@ -22,28 +23,88 @@ endif()
...
+            )
+        endif()
+    endif()
+
+    if(NOT DEFINED pybind11_VERSION)
+        # Search for a CMake config, which should have been installed with 
+        # pybind11, and will provide version information.
+        find_package(pybind11 ${pybind11_FIND_VERSION} CONFIG)
+
+        if(NOT pybind11_FOUND)
```
Comment (hodoulp) [inline share/cmake/modules/Findpybind11.cmake:56]: Even if an installed version is found, the version must still be validated.

Comment (michdolan) [reply share/cmake/modules/Findpybind11.cmake:56]: `find_package_handle_standard_args` will validate the version based on the `find_package` call that resolved to run this module.

File: share/cmake/modules/Findpybind11.cmake (hunk 1 above)
Comment (michdolan) [reply share/cmake/modules/Findpybind11.cmake:42]: Good find. I should have looked in more of the headers! I will get this incorporated.

Comment (michdolan) [reply share/cmake/modules/Findpybind11.cmake]: Added.

**Pull request merged**